02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`sha256`:`prevtx_txid`]
+    * [`u32`:`prevtx_vout`]
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+The sending node:
+  - MUST verify it has received valid commitment signatures from its peer
+  - MUST set `witness` to the serialized witness data for the input
+    corresponding to `prevtx_txid`:`prevtx_vout`
+  - MUST remember the details of this funding transaction.
+  - MUST NOT send a `witness_stack` whose length exceeds its
+    previously indicated `max_witness_len`.
+  - if it has NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `funding_signed2` message.
+
+The receiving node:
+  - if the `witness_stack` length exceeds `max_witness_len`:
+    - MUST error.
+  - MUST verify it has received valid commitment signatures
+  - SHOULD apply `witness`es to the funding transaction and
+    broadcast it.
+
+#### Rationale
+
+Every node must transmit their complete set of witness data for
+the inputs included in the funding transaction.
+
+`prevtx_txid` and `prevtx_vout` are the identifier for the input
+which the witness data pertain to
+
+`witness` is the data for a witness element in a witness stack
+
+### Kicking Off Replace-By-Fee: `init_rbf`
+
+This message initiates the flow to create a replacement funding
+transaction.  It is sent by the channel opener, after the funding
+transaction has been broadcast but before a `funding_locked` message
+is exchanged.
+
+Once an `init_rbf` message has been successfully ack'd by the accepter
+node, the message flow returns to `commitment_signed` and proceeds as
+indicated above.
+
+1. type: 62 (`init_rbf`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`funding_satoshis`]
+    * [`u32`:`feerate_per_kw`]
+    * [`u32`:`feerate_per_kw_funding`]
+    * [`u16`:`num_additional_inputs`]
+    * [`num_additional_inputs*input_info`:`input_info`]
+    * [`u16`:`num_outputs`]
+    * [`num_outputs*output_info`:`output_info`]
+
+#### Requirements
+
+The sending node:
+  - MUST have sent `open_channel2`.
+  - MUST send a `feerate_per_kw_funding` greater than the most recently
+    negotiated rate.
+  - MAY update the `feerate_per_kw`, the commitment transaction feerate.
+  - MAY include additional inputs.
+  - MAY set the `num_inputs` to zero.
+  - MUST transmit all outputs, excluduing the channel funding output.
+  - MAY include at least one output with the `sats` set to zero,
+    to be used as the change output.
+
+The receiving node:
+  - MUST return an error if:
+    - the `feerate_per_kw_funding` is not greater than the previously
+      negotiated rate.
+  - MUST return an error, but not fail the channel if:
```

Comment:
How does the remote node know that it should not fail the channel upon receiving the error ?

Reply:
This is, at best, a case for the 'soft error' idea that @rustyrussell has been talking about. Clearly the node that sent you the `init_rbf` is confused, the polite thing to do in the case is to inform them that they've made a mistake. The other option, without a 'soft error' is probably to ignore the `init_rbf` request entirely.

You make a good point though, what we really want here is that a node MUST NOT send an `init_rbf` if it has already sent a `funding_locked` message.